### DISCUSSION TITLE
`unigen` fails to build on a cross target due to libc linkage

### DISCUSSION THREAD

**Discussion by tranzystorekk** [ghostty-org/ghostty#3218] [Feature Requests, Ideas] [answered]
When testing `ghostty` 1.0.0 for Void Linux, building for a cross target currently fails.

We managed to identify the culprit as `unigen` using the `c_allocator` and linking to libc. It seems zig blindly follows the libc configuration it receives, and attempts to use it even for binaries targeted for host architecture.

This patch succeeds to work around the issue:

```diff
diff --git a/build.zig b/build.zig
index c3f7302..9fa73e9 100644
--- a/build.zig
+++ b/build.zig
@@ -1494,7 +1494,7 @@ fn addUnicodeTables(
             .root_source_file = b.path("src/unicode/props.zig"),
             .target = b.host,
         });
-        exe.linkLibC();
+        // exe.linkLibC();
         if (step_ == null) b.installArtifact(exe);
 
         const ziglyph_dep = b.dependency("ziglyph", .{
diff --git a/src/unicode/props.zig b/src/unicode/props.zig
index d77bf4c..994cf36 100644
--- a/src/unicode/props.zig
+++ b/src/unicode/props.zig
@@ -131,7 +131,7 @@ pub fn get(cp: u21) Properties {
 
 /// Runnable binary to generate the lookup tables and output to stdout.
 pub fn main() !void {
-    const alloc = std.heap.c_allocator;
+    const alloc = std.heap.page_allocator;
 
     const gen: lut.Generator(
         Properties,
```

I'm interested if it's acceptable, or if there's a reason for opting for system allocator here?

**Comment by rhodes-b** [comment]
I believe libc is a required library for the macos build as it interfaces with ghostty through exported c bindings

**Reply by tranzystorekk** [reply]
The `unigen` utility or ghostty in general?

**Reply by rhodes-b** [reply]
not sure what the `unigen` part is.

Swift cannot call zig code, so ghostty exports some C code so the macos app runtime can work properly

**Comment by morph027** [comment]
Same happens to me when building natively on Ubuntu 24.04. The patch indeed does work.

**Reply by morph027** [reply]
Also it struggles to find libadwaita and X11 headers... (not using snap zig install)

```
ln -s /usr/include/{libadwaita-1,X11} /tmp/zig-linux-x86-0.13.0/lib/include/
````

With installation from snap, it somehow finds the headers :shrug:

**Reply by rhodes-b** [reply]
> Also it struggles to find libadwaita and X11 headers... (not using snap zig install)
> 
> ```
> ln -s /usr/include/{libadwaita-1,X11} /tmp/zig-linux-x86-0.13.0/lib/include/
> ```
> 
> With installation from snap, it somehow finds the headers 🤷

You probably need `pkg-config` and the developer verisions of the libs like https://ghostty.org/docs/install/build#linux-installation-tips
mentions

**Reply by morph027** [reply]
I have them, it just does not find it in it's default location, that's why the link helps.

**Comment by squeek502** [comment, marked as answer]
No reason to use `c_allocator` here; just submitted a PR to replace it with ArenaAllocator: https://github.com/ghostty-org/ghostty/pull/5268
